src/minimega/router.go

```diff
@@ -0,0 +1,633 @@
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	log "minilog"
+	"net"
+	"path/filepath"
+	"ron"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"text/tabwriter"
+)
+
+var (
+	routers    map[int]*Router
+	routerLock sync.Mutex
+)
+
+type Router struct {
+	vmID         int        // local (and effectively unique regardless of namespace) vm id
+	IPs          [][]string // positional ip address (index 0 is the first listed network in vm config net)
+	lock         sync.Mutex
+	logLevel     string
+	updateIPs    bool // only update IPs if we've made changes
+	dhcp         map[string]*dhcp
+	dns          map[string]string
+	rad          map[string]bool // using a bool placeholder here for later RAD options
+	staticRoutes map[string]string
+	ospfRoutes   map[string]*ospf
+}
+
+type ospf struct {
+	area       string
+	interfaces map[string]bool
+}
+
+type dhcp struct {
+	addr   string
+	low    string
+	high   string
+	router string
+	dns    string
+	static map[string]string
+}
+
+func (r *Router) String() string {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// create output
+	var o bytes.Buffer
+	fmt.Fprintf(&o, "IPs:\n")
+	for i, v := range r.IPs {
+		fmt.Fprintf(&o, "Network: %v: %v\n", i, v)
+	}
+	fmt.Fprintln(&o)
+
+	if len(r.dhcp) > 0 {
+		var keys []string
+		for k, _ := range r.dhcp {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, k := range keys {
+			d := r.dhcp[k]
+			fmt.Fprintf(&o, "%v\n", d)
+		}
+	}
+
+	if len(r.dns) > 0 {
+		fmt.Fprintf(&o, "DNS:\n")
+		var keys []string
+		for k, _ := range r.dns {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, ip := range keys {
+			host := r.dns[ip]
+			fmt.Fprintf(&o, "%v\t%v\n", ip, host)
+		}
+		fmt.Fprintln(&o)
+	}
+
+	if len(r.rad) > 0 {
+		fmt.Fprintf(&o, "Router Advertisements:\n")
+		var keys []string
+		for k, _ := range r.rad {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, subnet := range keys {
+			fmt.Fprintf(&o, "%v\n", subnet)
+		}
+		fmt.Fprintln(&o)
+	}
+
+	if len(r.staticRoutes) > 0 {
+		fmt.Fprintf(&o, "Static Routes:\n")
+		var keys []string
+		for k, _ := range r.staticRoutes {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, network := range keys {
+			fmt.Fprintf(&o, "%v\t%v\n", network, r.staticRoutes[network])
+		}
+		fmt.Fprintln(&o)
+	}
+
+	if len(r.ospfRoutes) > 0 {
+		var keys []string
+		for k, _ := range r.ospfRoutes {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, k := range keys {
+			ospfRoute := r.ospfRoutes[k]
+			fmt.Fprintf(&o, "%v\n", ospfRoute)
+		}
+	}
+
+	vm := vms[r.vmID]
+	if vm == nil { // this really shouldn't ever happen
+		log.Error("could not find vm: %v", r.vmID)
+		return ""
+	}
+
+	lines := strings.Split(vm.Tag("minirouter_log"), "\n")
+
+	fmt.Fprintln(&o, "Log:")
+	for _, v := range lines {
+		fmt.Fprintf(&o, "\t%v\n", v)
+	}
+
+	return o.String()
+}
+
+func init() {
+	routers = make(map[int]*Router)
+}
```

Comment:
You don't have to do this anymore. I'd kill this and make it when you declare the var.